Realm/RLMChangeListener.mm

```diff
@@ -0,0 +1,156 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2014 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMChangeListener.h"
+
+#import "RLMRealm_Private.hpp"
+
+#include <tightdb/group_shared.hpp>
+#include <tightdb/commit_log.hpp>
+#include <tightdb/lang_bind_helper.hpp>
+
+// A weak holder for an RLMRealm to allow calling performSelector:onThread: without
+// a strong reference to the realm
+@interface RLMWeakNotifier : NSObject
+@end
+
+@implementation RLMWeakNotifier {
+    __weak RLMRealm *_realm;
+    // flag used to avoid queuing up redundant notifications
+    std::atomic_flag _hasPendingNotification;
+}
+
+- (instancetype)initWithRealm:(RLMRealm *)realm
+{
+    self = [super init];
+    if (self) {
+        _realm = realm;
+        _hasPendingNotification.clear();
+    }
+    return self;
+}
+
+- (void)notify
+{
+    _hasPendingNotification.clear();
+    [_realm handleExternalCommit];
+}
+
+- (void)notifyOnTargetThread
+{
+    RLMRealm *realm = _realm;
+    if (realm && !_hasPendingNotification.test_and_set()) {
+        [self performSelector:@selector(notify)
+                     onThread:realm->_thread withObject:nil waitUntilDone:NO];
+    }
+}
+@end
+
+@implementation RLMChangeListener {
+    NSMutableDictionary *_cache;
+    std::unique_ptr<Replication> _replication;
+    std::unique_ptr<SharedGroup> _sharedGroup;
+    NSString *_path;
+    NSMutableArray *_realms;
+    NSCondition *_shutdownCondition;
+}
+
+- (instancetype)initWithPath:(NSString *)path inMemory:(BOOL)inMemory cache:(NSMutableDictionary *)cache {
+    self = [super initWithTarget:self selector:@selector(run) object:nil];
+    if (self) {
+        _cache = cache;
+        _replication.reset(tightdb::makeWriteLogCollector(path.UTF8String));
+        SharedGroup::DurabilityLevel durability = inMemory ? SharedGroup::durability_MemOnly :
+                                                             SharedGroup::durability_Full;
+        _sharedGroup = std::make_unique<SharedGroup>(*_replication, durability);
+        _sharedGroup->begin_read();
+
+        _path = path;
+        _realms = [NSMutableArray array];
+        _shutdownCondition = [[NSCondition alloc] init];
+        [self start];
+    }
+    return self;
+}
+
+- (void)addRealm:(RLMRealm *)realm {
+    @synchronized (_realms) {
```

Comment:
Would be good to only synchronize on _cache (or the global once it is moved into this file).